File: packages/sdk-go/evaluator.go
package agentshield

import (
	"encoding/json"
	"fmt"
	"regexp"
	"sync"
	"time"
)

type PolicyEvaluator struct {
	callHistory map[string][]int64
	sessionCost map[string]float64
	circuitBreaker *CircuitBreaker
	mu sync.RWMutex
}

func NewPolicyEvaluator() *PolicyEvaluator {
	return &PolicyEvaluator{
		callHistory:    make(map[string][]int64),
		sessionCost:    make(map[string]float64),
		circuitBreaker: NewCircuitBreaker(),
	}
}

func (e *PolicyEvaluator) Evaluate(request ToolCallRequest, policy GuardrailPolicy) EvaluationResult {
	timestamp := time.Now().UTC().Format(time.RFC3339)
	sessionKey := request.SessionID
	if sessionKey == "" {
		sessionKey = request.AgentID
	}
	if sessionKey == "" {
		sessionKey = "default-session"
	}

	if policy.CircuitBreaker != nil {
		cbCheck := e.circuitBreaker.Check(sessionKey, request.ToolName, request.Params, *policy.CircuitBreaker)
		if cbCheck.Tripped {
			return EvaluationResult{
				Allowed:     false,
				Reason:      cbCheck.Reason,
				ActionTaken: ActionCircuitTripped,
				Timestamp:   timestamp,
				Remediation: &Remediation{
					Status:       "BLOCKED",
					SuggestedFix: "Agent in retry loop. Abort current tool sequence and ask human for clarification.",
				},
			}
		}
	}

	if policy.EnableInjectionSanitizer != false {
		payloadStr, _ := json.Marshal(request.Params)
		sanitizeRes := InspectInjection(string(payloadStr))
		if sanitizeRes.Detected {
			return EvaluationResult{
				Allowed:     false,
				Reason:      fmt.Sprintf("Security Threat Detected: %s (%s).", sanitizeRes.Type, sanitizeRes.PatternMatched),
				ActionTaken: ActionBlock,
				Timestamp:   timestamp,
				Remediation: &Remediation{
					Status:       "BLOCKED",
					SuggestedFix: "Sanitize input payload to remove prompt overrides or hidden unicode characters.",
				},
			}
		}
	}

	if len(policy.AllowedTools) > 0 {
		allowed := false
		for _, t := range policy.AllowedTools {
			if t == request.ToolName {
				allowed = true
				break
			}
		}
		if !allowed {
			return EvaluationResult{
				Allowed:     false,
				Reason:      fmt.Sprintf("Tool '%s' is not in the allowed tools list.", request.ToolName),
				ActionTaken: ActionBlock,
				Timestamp:   timestamp,
				Remediation: &Remediation{
					Status:       "BLOCKED",
					SuggestedFix: fmt.Sprintf("Tool '%s' is not authorized. Permitted tools: %v.", request.ToolName, policy.AllowedTools),
				},
			}
		}
	}

	for _, t := range policy.ForbiddenTools {
		if t == request.ToolName {
			return EvaluationResult{
				Allowed:     false,
				Reason:      fmt.Sprintf("Tool '%s' is explicitly forbidden by policy.", request.ToolName),
				ActionTaken: ActionBlock,
				Timestamp:   timestamp,
				Remediation: &Remediation{
					Status:       "BLOCKED",
					SuggestedFix: fmt.Sprintf("Tool '%s' is prohibited in production environment.", request.ToolName),
				},
			}
		}
	}

	for _, field := range policy.RequiredFields {
		if _, ok := request.Params[field]; !ok {
			return EvaluationResult{
				Allowed:     false,
				Reason:      fmt.Sprintf("Missing required parameter field '%s'.", field),
				ActionTaken: ActionBlock,
				Timestamp:   timestamp,
				Remediation: &Remediation{
					Status:       "REQUIRES_REMEDIATION",
					SuggestedFix: fmt.Sprintf("Provide required parameter '%s' before invoking tool '%s'.", field, request.ToolName),
				},
			}
		}
	}

	if policy.RateLimit != nil {
		e.mu.Lock()
		now := time.Now().UnixMilli()
		oneMinuteAgo := now - 60000

		timestamps := e.callHistory[sessionKey]
		var recent []int64
		for _, t := range timestamps {
			if t > oneMinuteAgo {
				recent = append(recent, t)
			}
		}

		if len(recent) >= policy.RateLimit.MaxCallsPerMinute {
			e.mu.Unlock()
			return EvaluationResult{
				Allowed:     false,
				Reason:      fmt.Sprintf("Rate limit exceeded: Max %d calls/min allowed.", policy.RateLimit.MaxCallsPerMinute),
				ActionTaken: ActionBlock,
				Timestamp:   timestamp,
				Remediation: &Remediation{
					Status:       "BLOCKED",
					SuggestedFix: fmt.Sprintf("Wait 60 seconds before issuing further tool calls for session '%s'.", sessionKey),
				},
			}
		}

		recent = append(recent, now)
		e.callHistory[sessionKey] = recent
		e.mu.Unlock()
	}

	if policy.MaxCostPerSession > 0 && request.EstimatedCost > 0 {
		e.mu.Lock()
		currentCost := e.sessionCost[sessionKey]
		newCost := currentCost + request.EstimatedCost

		if newCost > policy.MaxCostPerSession {
			e.mu.Unlock()
			return EvaluationResult{
				Allowed:     false,
				Reason:      fmt.Sprintf("Session cost threshold exceeded ($%.4f > $%.4f cap).", newCost, policy.MaxCostPerSession),
				ActionTaken: ActionBlock,
				Timestamp:   timestamp,
				Remediation: &Remediation{
					Status:       "BLOCKED",
					SuggestedFix: fmt.Sprintf("Budget limit reached ($%.2f). Request budget approval.", policy.MaxCostPerSession),
				},
			}
		}
		e.sessionCost[sessionKey] = newCost
		e.mu.Unlock()
	}

	for paramKey, maxValue := range policy.MaxParamValues {
		if actualVal, ok := request.Params[paramKey]; ok {
			var fv float64
			switch v := actualVal.(type) {
			case float64:
				fv = v
			case int:
				fv = float64(v)
			case int64:
				fv = float64(v)
			case json.Number:
				fv, _ = v.Float64()
			default:
				continue
			}
			if fv > maxValue {
				return EvaluationResult{
					Allowed:     false,
					Reason:      fmt.Sprintf("Parameter '%s' value (%.0f) exceeds maximum allowed threshold (%.0f).", paramKey, fv, maxValue),
					ActionTaken: ActionBlock,
					Timestamp:   timestamp,
					Remediation: &Remediation{
						Status:          "REQUIRES_REMEDIATION",
						SuggestedFix:    fmt.Sprintf("Reduce '%s' parameter to <= %.0f.", paramKey, maxValue),
						MaxAllowedValue: maxValue,
					},
				}
			}
		}
	}

	payloadStr, _ := json.Marshal(request.Params)
	for _, patternStr := range policy.ForbiddenPatterns {
		matched, _ := regexp.MatchString("(?i)"+patternStr, string(payloadStr))
		if matched {
			return EvaluationResult{
				Allowed:     false,
				Reason:      fmt.Sprintf("Parameter payload matched forbidden pattern: '%s'.", patternStr),
				ActionTaken: ActionBlock,
				Timestamp:   timestamp,
				Remediation: &Remediation{
					Status:       "BLOCKED",
					SuggestedFix: fmt.Sprintf("Remove forbidden pattern '%s' from input payload.", patternStr),
				},
			}
		}
	}

	if policy.RequireApproval {
		return EvaluationResult{
			Allowed:     false,
			Reason:      fmt.Sprintf("Tool '%s' requires human authorization prior to execution.", request.ToolName),
			ActionTaken: ActionRequireApproval,
			Timestamp:   timestamp,
			Remediation: &Remediation{
				Status:       "BLOCKED",
				SuggestedFix: "Request human authorization token.",
			},
		}
	}

	return EvaluationResult{
		Allowed:     true,
		ActionTaken: ActionAllow,
		Timestamp:   timestamp,
	}
}

File: packages/sdk-rust/src/evaluator.rs
use dashmap::DashMap;
use regex::Regex;
use serde_json::{json, Value};
use sha2::{Sha256, Digest};
use std::collections::HashMap;
use std::sync::Arc;

pub struct PolicyEvaluator {
    call_history: Arc<DashMap<String, Vec<u64>>>,
    session_costs: Arc<DashMap<String, f64>>,
    circuit_breaker: crate::circuit_breaker::CircuitBreaker,
}

impl Default for PolicyEvaluator {
    fn default() -> Self {
        Self::new()
    }
}

impl PolicyEvaluator {
    pub fn new() -> Self {
        Self {
            call_history: Arc::new(DashMap::new()),
            session_costs: Arc::new(DashMap::new()),
            circuit_breaker: crate::circuit_breaker::CircuitBreaker::new(),
        }
    }

    pub fn evaluate(&self, request: &crate::types::ToolCallRequest, policy: &crate::types::GuardrailPolicy) -> crate::types::EvaluationResult {
        let timestamp = chrono::Utc::now().to_rfc3339();
        let session_key = request.session_id.as_deref()
            .or(request.agent_id.as_deref())
            .unwrap_or("default-session");

        if let Some(cb_config) = &policy.circuit_breaker {
            let cb_check = self.circuit_breaker.check(session_key, &request.tool_name, &request.params, cb_config);
            if cb_check.tripped {
                return crate::types::EvaluationResult {
                    allowed: false,
                    reason: cb_check.reason,
                    action_taken: crate::types::ActionTaken::CircuitTripped,
                    timestamp,
                    remediation: Some(crate::types::Remediation {
                        status: "BLOCKED".to_string(),
                        suggested_fix: Some("Agent in retry loop. Abort current tool sequence and ask human for clarification.".to_string()),
                        max_allowed_value: None,
                    }),
                };
            }
        }

        if policy.enable_injection_sanitizer.unwrap_or(true) {
            let payload_str = serde_json::to_string(&request.params).unwrap_or_default();
            let sanitize_res = crate::sanitizer::InjectionSanitizer::inspect(&payload_str);
            if sanitize_res.detected {
                return crate::types::EvaluationResult {
                    allowed: false,
                    reason: Some(format!("Security Threat Detected: {} ({}).", sanitize_res.r#type.unwrap_or_default(), sanitize_res.pattern_matched.unwrap_or_default())),
                    action_taken: crate::types::ActionTaken::Block,
                    timestamp,
                    remediation: Some(crate::types::Remediation {
                        status: "BLOCKED".to_string(),
                        suggested_fix: Some("Sanitize input payload to remove prompt overrides or hidden unicode characters.".to_string()),
                        max_allowed_value: None,
                    }),
                };
            }
        }

        if let Some(allowed) = &policy.allowed_tools {
            if !allowed.is_empty() && !allowed.contains(&request.tool_name) {
                return crate::types::EvaluationResult {
                    allowed: false,
                    reason: Some(format!("Tool '{}' is not in the allowed tools list.", request.tool_name)),
                    action_taken: crate::types::ActionTaken::Block,
                    timestamp,
                    remediation: Some(crate::types::Remediation {
                        status: "BLOCKED".to_string(),
                        suggested_fix: Some(format!("Tool '{}' is not authorized. Permitted tools: {}.", request.tool_name, allowed.join(", "))),
                        max_allowed_value: None,
                    }),
                };
            }
        }

        if let Some(forbidden) = &policy.forbidden_tools {
            if forbidden.contains(&request.tool_name) {
                return crate::types::EvaluationResult {
                    allowed: false,
                    reason: Some(format!("Tool '{}' is explicitly forbidden by policy.", request.tool_name)),
                    action_taken: crate::types::ActionTaken::Block,
                    timestamp,
                    remediation: Some(crate::types::Remediation {
                        status: "BLOCKED".to_string(),
                        suggested_fix: Some(format!("Tool '{}' is prohibited in production environment.", request.tool_name)),
                        max_allowed_value: None,
                    }),
                };
            }
        }

        if let Some(required) = &policy.required_fields {
            for field in required {
                if !request.params.contains_key(field) || request.params[field].is_null() {
                    return crate::types::EvaluationResult {
                        allowed: false,
                        reason: Some(format!("Missing required parameter field '{}'.", field)),
                        action_taken: crate::types::ActionTaken::Block,
                        timestamp,
                        remediation: Some(crate::types::Remediation {
                            status: "REQUIRES_REMEDIATION".to_string(),
                            suggested_fix: Some(format!("Provide required parameter '{}' before invoking tool '{}'.", field, request.tool_name)),
                            max_allowed_value: None,
                        }),
                    };
                }
            }
        }

        if let Some(rate_limit) = &policy.rate_limit {
            let now = chrono::Utc::now().timestamp_millis() as u64;
            let one_minute_ago = now.saturating_sub(60000);

            let mut timestamps = self.call_history.entry(session_key.to_string()).or_insert_with(Vec::new);
            timestamps.retain(|&t| t > one_minute_ago);

            if timestamps.len() >= rate_limit.max_calls_per_minute as usize {
                return crate::types::EvaluationResult {
                    allowed: false,
                    reason: Some(format!("Rate limit exceeded: Max {} calls/min allowed.", rate_limit.max_calls_per_minute)),
                    action_taken: crate::types::ActionTaken::Block,
                    timestamp,
                    remediation: Some(crate::types::Remediation {
                        status: "BLOCKED".to_string(),
                        suggested_fix: Some(format!("Wait 60 seconds before issuing further tool calls for session '{}'.", session_key)),
                        max_allowed_value: None,
                    }),
                };
            }

            timestamps.push(now);
        }

        if let Some(max_cost) = policy.max_cost_per_session {
            if let Some(estimated) = request.estimated_cost {
                let mut current_cost = self.session_costs.entry(session_key.to_string()).or_insert(0.0);
                let new_cost = *current_cost + estimated;

                if new_cost > max_cost {
                    return crate::types::EvaluationResult {
                        allowed: false,
                        reason: Some(format!("Session cost threshold exceeded (${:.4} > ${:.4} cap).", new_cost, max_cost)),
                        action_taken: crate::types::ActionTaken::Block,
                        timestamp,
                        remediation: Some(crate::types::Remediation {
                            status: "BLOCKED".to_string(),
                            suggested_fix: Some(format!("Budget limit reached (${:.2}). Request budget approval.", max_cost)),
                            max_allowed_value: None,
                        }),
                    };
                }
                *current_cost = new_cost;
            }
        }

        if let Some(max_params) = &policy.max_param_values {
            for (param_key, max_value) in max_params {
                if let Some(Value::Number(actual)) = request.params.get(param_key) {
                    if let Some(fv) = actual.as_f64() {
                        if fv > *max_value {
                            return crate::types::EvaluationResult {
                                allowed: false,
                                reason: Some(format!("Parameter '{}' value ({}) exceeds maximum allowed threshold ({}).", param_key, fv, max_value)),
                                action_taken: crate::types::ActionTaken::Block,
                                timestamp,
                                remediation: Some(crate::types::Remediation {
                                    status: "REQUIRES_REMEDIATION".to_string(),
                                    suggested_fix: Some(format!("Reduce '{}' parameter to <= {}.", param_key, max_value)),
                                    max_allowed_value: Some(*max_value),
                                }),
                            };
                        }
                    }
                }
            }
        }

        if let Some(patterns) = &policy.forbidden_patterns {
            let param_str = serde_json::to_string(&request.params).unwrap_or_default();
            for pattern_str in patterns {
                if Regex::new(&format!("(?i){}", pattern_str)).unwrap().is_match(&param_str) {
                    return crate::types::EvaluationResult {
                        allowed: false,
                        reason: Some(format!("Parameter payload matched forbidden pattern: '{}'.", pattern_str)),
                        action_taken: crate::types::ActionTaken::Block,
                        timestamp,
                        remediation: Some(crate::types::Remediation {
                            status: "BLOCKED".to_string(),
                            suggested_fix: Some(format!("Remove forbidden pattern '{}' from input payload.", pattern_str)),
                            max_allowed_value: None,
                        }),
                    };
                }
            }
        }

        if policy.require_approval.unwrap_or(false) {
            return crate::types::EvaluationResult {
                allowed: false,
                reason: Some(format!("Tool '{}' requires human authorization prior to execution.", request.tool_name)),
                action_taken: crate::types::ActionTaken::RequireApproval,
                timestamp,
                remediation: Some(crate::types::Remediation {
                    status: "BLOCKED".to_string(),
                    suggested_fix: Some("Request human authorization token.".to_string()),
                    max_allowed_value: None,
                }),
            };
        }

        crate::types::EvaluationResult {
            allowed: true,
            reason: None,
            action_taken: crate::types::ActionTaken::Allow,
            timestamp,
            remediation: None,
        }
    }
}